src/fromager/bootstrapper.py

```diff
@@ -0,0 +1,489 @@
+from __future__ import annotations
+
+import logging
+import operator
+import pathlib
+import shutil
+import typing
+
+from packaging.requirements import Requirement
+from packaging.utils import canonicalize_name
+from packaging.version import Version
+
+from . import (
+    build_environment,
+    dependencies,
+    finders,
+    progress,
+    resolver,
+    server,
+    sources,
+    wheels,
+)
+from .dependency_graph import DependencyGraph
+from .requirements_file import RequirementType
+
+if typing.TYPE_CHECKING:
+    from . import context
+
+logger = logging.getLogger(__name__)
+
+
+class Bootstrapper:
+    def __init__(
+        self,
+        ctx: context.WorkContext,
+        progressbar: progress.Progressbar | None = None,
+        prev_graph: DependencyGraph | None = None,
+    ) -> None:
+        self.ctx = ctx
+        self.progressbar = progressbar or progress.Progressbar(None)
+        self.prev_graph = prev_graph
+        self.why: list[tuple[RequirementType, Requirement, Version]] = []
+
+    def bootstrap(self, req: Requirement, req_type: RequirementType) -> Version:
+        constraint = self.ctx.constraints.get_constraint(req.name)
+        if constraint:
+            logger.info(
+                f"{req.name}: incoming requirement {req} matches constraint {constraint}. Will apply both."
+            )
+
+        pbi = self.ctx.package_build_info(req)
+        if pbi.pre_built:
+            resolved_version, wheel_url, wheel_filename, unpack_dir = (
+                self._handle_prebuilt_sources(req, req_type)
+            )
+            # Remember that this is a prebuilt wheel, and where we got it.
+            source_url = wheel_url
+            source_url_type = "prebuilt"
```

Comment:
Suggestion: add "prebuilt" to RequirementType enum

Reply:
This is more for sdist types which are: "sdist", "override" and "prebuilt". I can create a new enum for that